dashboard/config/routes.rb

```diff
@@ -234,6 +234,7 @@
   resources :level_starter_assets, only: [:show], param: 'level_name' do
     member do
       get '/:filename', to: 'level_starter_assets#file'
+      post '', to: 'level_starter_assets#upload'
```

Comment:
if you rename your controller methods to `get` and `create`, you should be able to remove the `member do` block and just list these in `resources :level_starter_assets, only: [:show, :create]`. For example, the following resource definition:
```
  concern :section_api_routes do
    resources :sections, only: [:index, :show, :create, ...]
```
generates the following routes:
```
$ rails routes | grep /api/v1/sections
...
api_v1_sections GET      /api/v1/sections(.:format)                                                                   api/v1/sections#index
                POST     /api/v1/sections(.:format)                                                                   api/v1/sections#create
                GET      /api/v1/sections/:id(.:format)                                                               api/v1/sections#show
```

I may not have the syntax exactly right in my code snippet at the top, but it's worth figuring out the rails syntax / convention for this, and saving the `member` directive for when you really need to do something special.

Reply:
i think both of these routes need member entries for different but similar reasons:

- Rails convention for `:get` is `GET /level_starter_assets/:level_name`, but we need `GET /level_starter_assets/:level_name/:filename` in order to get the resource
- Rails convention for `:post` is `POST /level_starter_assets`, but we need `POST /level_starter_assets/:level_name` to create the resource

are there other/better ways to define these special cases rather than by using members?

Reply:
something similar is possible based on the following example:
```ruby
  resources :scripts, path: '/s/' do
    # /s/xxx/stage/yyy/puzzle/zzz
    resources :stages, only: [], path: "/stage", param: 'position', format: false do
      resources :script_levels, only: [:show], path: "/puzzle", format: false do
```
that example ends up being slightly different because it uses extra path parts (`/foo/:foo_id/bar/:bar_id` instead of /foo/:foo_id/:bar_id`). it may be possible to eliminate those, but at that point it'll probably get more complicated than what you have here.

One idea would be to use the routes:
```
POST /level_names/:level_name/starter_assets
GET /level_names/:level_name/starter_assets/:id
DELETE /level_names/:level_name/starter_assets/:id
```
which you could get by doing
```ruby
resources :level_names, only: [] do
  resources :starter_assets, only: [:create, :show, :destroy]
end
```
this is a bit more railsy but still a a bit weird because you are accessing the level by name whereas the rails way would be to access it by id.

its fine if you want to punt on this but worth noting that it'll be harder to change once external users start relying on the GET path (the POST and DELETE aren't too bad because levelbuilder updates are a bit more atomic and so you don't really get the deploy problems described in https://github.com/code-dot-org/code-dot-org/pull/30345)

Reply:
yeah, i think i will punt on this because the end goal is to totally remove the GET path in favor of refactoring the client to rely on an S3 URL rather than endpoints on our server that read from S3